// ng-zzfx/esm2015/lib/zoundfx.directive.js
import { __decorate } from "tslib";
import { AfterViewInit, Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import * as zzfx from './source/zzfx';
let ZoundfxDirective = class ZoundfxDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.vol = 0.1; // Number from 0 to 1;
        this.trigger = 'onclick';
    }
    ngAfterViewInit() {
        this.zzfxFn = zzfx.ZZFX(this.vol);
        this.registerEventListeners();
    }
    registerEventListeners() {
        this.renderer.listen(this.el.nativeElement, this.trigger, () => {
            this.zzfxFn(...this.zound);
        });
    }
};
ZoundfxDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input()
], ZoundfxDirective.prototype, "vol", void 0);
__decorate([
    Input()
], ZoundfxDirective.prototype, "trigger", void 0);
__decorate([
    Input()
], ZoundfxDirective.prototype, "zound", void 0);
ZoundfxDirective = __decorate([
    Directive({
        selector: '[zoundfx]'
    })
], ZoundfxDirective);
export { ZoundfxDirective };
