// ng-zzfx/esm5/lib/zoundfx.directive.js
import { __decorate, __read, __spread } from "tslib";
import { AfterViewInit, Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import * as zzfx from './source/zzfx';
var ZoundfxDirective = /** @class */ (function () {
    function ZoundfxDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.vol = 0.1; // Number from 0 to 1;
        this.trigger = 'onclick';
    }
    ZoundfxDirective.prototype.ngAfterViewInit = function () {
        this.zzfxFn = zzfx.ZZFX(this.vol);
        this.registerEventListeners();
    };
    ZoundfxDirective.prototype.registerEventListeners = function () {
        var _this = this;
        this.renderer.listen(this.el.nativeElement, this.trigger, function () {
            _this.zzfxFn.apply(_this, __spread(_this.zound));
        });
    };
    ZoundfxDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input()
    ], ZoundfxDirective.prototype, "vol", void 0);
    __decorate([
        Input()
    ], ZoundfxDirective.prototype, "trigger", void 0);
    __decorate([
        Input()
    ], ZoundfxDirective.prototype, "zound", void 0);
    ZoundfxDirective = __decorate([
        Directive({
            selector: '[zoundfx]'
        })
    ], ZoundfxDirective);
    return ZoundfxDirective;
}());
export { ZoundfxDirective };
